### ISSUE TITLE
What is "nn.BatchNorm2d" for?

### ISSUE THREAD

**Issue by ghbcpeng** [2017-fall-DL-training-program/ConvNetwork#3]
Dear TA,

In the example code from GitHub https://github.com/kuangliu/pytorch-cifar, _nn.BatchNorm2d_ is invoked in file resnet.py to build _BasicBlock_. This function seems not shown in Figure 2 in the lab handout, and the images are already normalized by _trainloader_ and _testloader_. Then why do we need _nn.BatchNorm2d_? Thanks for your reply.

Regards,
BC

**Comment by JiaRenChang** [comment]
Hi,
   Yes, you need batch normalization.
   The only one reason why the Figure 2 do not draw Batch Normalization (**BN**) layer is for **simplification**.
   Because BN is very important to train deep networks.
   If you want to know why BN is essential to deep learning, you can refer to this paper "[Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift](http://machinelearning.wustl.edu/mlpapers/paper_files/icml2015_ioffe15.pdf) ".

Jia-Ren

**Comment by bcpenggh** [comment]
Understood. Thanks

**Comment by jessejchuang** [comment]
Hi TA,

In Pytorch BN module, we can see four parameters as below.
**_>>> self.bn1.state_dict().keys()
... odict_keys(['weight', 'bias', 'running_mean', 'running_var'])_**

I know **‘running_mean’** and **‘running_var’** are **beta** and **gamma** respectively and they are **learnable parameters in training mode** (then fixed in evaluation mode).

But for **'weight'** and **'bias'**, they are **inherited from parent module** but **seem useless in batchNorm calculation**. Do I misunderstand weight/bias? Any comments?

**Comment by JiaRenChang** [comment]
Hi,
You misunderstand.
The weight is **gamma** and the bias is **beta** according to the paper.
The running mean and variance is for inference.
You can refer to the **stage 10 in algorithm 2** in the paper.
Jia-Ren
